File: [1329][Medium][Sort_the_Matrix_Diagonally]/Sort_the_Matrix_Diagonally.cpp
// solution 3,
// leetcode time     cost : 12 ms
// leetcode memory   cost : 11 MB
// Time  Complexity: O(mn log(min(m,n))
// Space Complexity: O(min(m,n))
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();
        for (int i = 0; i < m; i++) {
            // combine sort(i,0) and sort(0,j) loop with 2 nested for loop and "i?1:n"
            for (int j = 0; j < (i?1:n); j++) {
                vector<int> vals;
                while (i<m && j<n)
                    vals.push_back( mat[i++][j++] );
                sort(vals.begin(), vals.end());
                while (i && j) {
                    mat[--i][--j] = vals.back();
                    vals.pop_back();
                }
            }
        }
        return mat;
    }
};

File: [1329][Medium][Sort_the_Matrix_Diagonally]/Sort_the_Matrix_Diagonally.java
// solution 3,
// leetcode time     cost : 11 ms
// leetcode memory   cost : 40.5 MB
// Time  Complexity: O(mn log(min(m,n))
// Space Complexity: O(min(m,n))
class Solution {
    public int[][] diagonalSort(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        for (int i = 0; i < m; i++) {
            // combine sort(i,0) and sort(0,j) loop with 2 nested for loop and "i>0?1:n"
            for (int j = 0; j < (i>0?1:n); j++) {
                Stack<Integer> vals = new Stack<>();
                while (i<m && j<n)
                    vals.add(mat[i++][j++]);
                Collections.sort(vals);
                while (i > 0 && j > 0)
                    mat[--i][--j] = vals.pop();
            }
        }
        return mat;
    }
}